doc/src/build/programming-with-objects/ch4-object-wrapping.md

```diff
@@ -0,0 +1,195 @@
+
+To be able to perform a swap of objects, both objects must be owned by the same account. We can imagine that a third-party builds infrastructure to provide swap serviecs. Anyone who wants to swap their object can send their objects to the third-party, and the third-party will help perform the swap and send them back. However, we don't fully trust the third-party and don't want to give them full custody of our objects. To achieve this, we can use direct wrapping. We define a wrapper object type as following:
+```rust
+struct ObjectWrapper has key {
+    id: VersionedID,
+    original_owner: address,
+    to_swap: Object,
+    fee: Balance<SUI>,
+}
+```
+`ObjectWrapper` defines a Sui object type, wraps the object that we want to swap as `to_swap`, and tracks the original owner of the object in `original_owner`. To make this more interesting and realistic, we can also expect that we may need to pay the third-party some fee for this swap. Below we define an interface to request a swap by someone who owns an `Object`:
+```rust
+public(script) fun request_swap(object: Object, fee: Coin<SUI>, service_address: address, ctx: &mut TxContext) {
+    assert!(Coin::value(&fee) >= MIN_FEE, 0);
+    let wrapper = ObjectWrapper {
+        id: TxContext::new_id(ctx),
+        original_owner: TxContext::sender(ctx),
+        to_swap: object,
+        fee: Coin::into_balance(fee),
+    };
+    Transfer::transfer(wrapper, service_address);
+}
+```
+In the above entry function, to request for swapping an `object`, one must pass the object by value so that it's fully consumed and wrapped into `ObjectWrapper`. A fee (in the type of `Coin<SUI>`) is also provided. It also checks that the fee is sufficient. Note that we turn `Coin` into `Balance` when putting it into the `wrapper` object. This is because `Coin` is a Sui object type and only used to pass around as Sui objects (e.g. as entry function arguments or objects sent to addresses). For coin balances that need to be embedded in another Sui object struct, we use `Balance` instead because it's not a Sui object type and hence is much cheaper to use.
+The wrapper object is then sent to the service operator, whose address is also specified in the call as `serviec_address`.
+
+Although the service operator (`service_address`) now owns the `ObjectWrapper`, which contains the object to be swapped, there is nothing they can do about the object. They cannot transfer it even though we have defined a `transfer_object` entry function for `Object`. This is because they cannot pass the wrapped `Object` as an argument to the function.
```

Comment:
```suggestion
Although the service operator (`service_address`) now owns the `ObjectWrapper`, which contains the object to be swapped, they cannot access the wrapped `Object. In particular, the operator cannot transfer it even though we have defined a `transfer_object` entry function for `Object`. This is because they cannot pass the wrapped `Object` as an argument to the function.
```